src/app/clusters/access-control-server/access-control-server.cpp

```diff
@@ -287,12 +288,99 @@ class AccessControlAttribute : public chip::app::AttributeAccessInterface
 private:
     CHIP_ERROR ReadAcl(AttributeValueEncoder & aEncoder);
     CHIP_ERROR ReadExtension(AttributeValueEncoder & aEncoder);
-    CHIP_ERROR WriteAcl(AttributeValueDecoder & aDecoder);
+    CHIP_ERROR WriteAcl(AttributeValueDecoder & aDecoder, EndpointId endpointId);
     CHIP_ERROR WriteExtension(AttributeValueDecoder & aDecoder);
 };
 
 constexpr uint16_t AccessControlAttribute::ClusterRevision;
 
+CHIP_ERROR GetAccessControlEntry(const AccessControl::Entry & entry,
+                                 DataModel::Nullable<AccessControlCluster::Structs::AccessControlEntry::Type> & accessControlEntry)
+{
+    AccessControlCluster::Structs::AccessControlEntry::Type staging;
+
+    ReturnErrorOnFailure(entry.GetFabricIndex(staging.fabricIndex));
+
+    {
+        Privilege privilege;
+        ReturnErrorOnFailure(entry.GetPrivilege(privilege));
+        ReturnErrorOnFailure(AccessControlEntryCodec::Convert(privilege, staging.privilege));
+    }
+
+    {
+        AuthMode authMode;
+        ReturnErrorOnFailure(entry.GetAuthMode(authMode));
+        ReturnErrorOnFailure(AccessControlEntryCodec::Convert(authMode, staging.authMode));
+    }
+
+    NodeId subjectBuffer[CHIP_CONFIG_EXAMPLE_ACCESS_CONTROL_MAX_SUBJECTS_PER_ENTRY];
+    size_t subjectCount;
+    ReturnErrorOnFailure(entry.GetSubjectCount(subjectCount));
+    if (subjectCount > 0)
+    {
+        for (size_t i = 0; i < subjectCount; ++i)
+        {
+            ReturnErrorOnFailure(entry.GetSubject(i, subjectBuffer[i]));
+        }
+        staging.subjects.SetNonNull(subjectBuffer, subjectCount);
+    }
+
+    AccessControlCluster::Structs::Target::Type targetBuffer[CHIP_CONFIG_EXAMPLE_ACCESS_CONTROL_MAX_TARGETS_PER_ENTRY];
+    size_t targetCount;
+    ReturnErrorOnFailure(entry.GetTargetCount(targetCount));
+    if (targetCount > 0)
+    {
+        for (size_t i = 0; i < targetCount; ++i)
+        {
+            AccessControl::Entry::Target target;
+            ReturnErrorOnFailure(entry.GetTarget(i, target));
+            ReturnErrorOnFailure(AccessControlEntryCodec::Convert(target, targetBuffer[i]));
+        }
+        staging.targets.SetNonNull(targetBuffer, targetCount);
+    }
+
+    accessControlEntry.SetNonNull(staging);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR LogAccessControlEvent(EndpointId endpointId, const AccessControl::Entry & entry,
+                                 const Access::SubjectDescriptor & subjectDescriptor,
+                                 AccessControlCluster::ChangeTypeEnum changeType)
+{
+    CHIP_ERROR ret;
```

Comment:
Have been using `err` all throughout access control.
```suggestion
    CHIP_ERROR err;
```